File: Android/Blink/libjava/src/main/java/net/qiujuer/blink/core/SendDispatcher.java
/*
 * Copyright (C) 2014 Qiujuer <qiujuer@live.cn>
 * WebSite http://www.qiujuer.net
 * Created 03/31/2015
 * Changed 04/02/2015
 * Version 1.0.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.qiujuer.blink.core;

import java.util.concurrent.BlockingQueue;

/**
 * Provides a thread for performing send dispatch from a queue of BinkConn {@link BlinkConn}.
 */
public class SendDispatcher extends Thread {
    /**
     * The queue of send entity.
     */
    private final BlockingQueue<SendPacket<?>> mQueue;
    /**
     * The sender interface for processing sender requests.
     */
    private final Sender mSender;
    /**
     * For posting send responses.
     */
    private final SendDelivery mDelivery;
    /**
     * Used for telling us to die.
     */
    private volatile boolean mQuit = false;

    public SendDispatcher(BlockingQueue<SendPacket<?>> queue,
                          Sender sender, SendDelivery delivery) {
        mQueue = queue;
        mSender = sender;
        mDelivery = delivery;
    }

    /**
     * Forces this dispatcher to quit immediately.  If any requests are still in
     * the queue, they are not guaranteed to be processed.
     */
    public void quit() {
        mQuit = true;
        interrupt();
    }

    @Override
    public void run() {
        while (true) {
            SendPacket<?> entity;
            try {
                // Take a request from the queue.
                entity = mQueue.take();
            } catch (InterruptedException e) {
                // We may have been interrupted because it was time to quit.
                if (mQuit) {
                    return;
                }
                continue;
            }

            try {
                if (entity.isCanceled()) {
                    continue;
                }
                // Post Start
                mDelivery.postSendStart(entity);

                // Send
                boolean status = mSender.sendHead(entity) && mSender.sendEntity(entity, mDelivery);

                // Post End
                mDelivery.postSendEnd(entity, status);

            } catch (Exception e) {
                e.printStackTrace();
                //mDelivery.postSendError();
            }
        }
    }
}

File: C Sharp/Blink/Blink/Async/SendDispatcher.cs
using Net.Qiujuer.Blink.Core;
using Net.Qiujuer.Blink.Listener.Delivery;
using System;
using System.Collections.Concurrent;
using System.Net.Sockets;

namespace Net.Qiujuer.Blink.Async
{
    /// <summary>
    /// Provides a thread for performing send dispatch from a queue of BinkConn {@link BlinkConn}.
    /// </summary>
    public class SendDispatcher : SocketAsyncEventArgs
    {
        private const int HeadSize = 11;
        /// <summary>
        /// The queue of send entity.
        /// </summary>
        private readonly ConcurrentQueue<SendPacket> mQueue;

        /// <summary>
        /// The sender interface for processing sender requests.
        /// </summary>
        private Sender mSender;

        /// <summary>
        /// Posting send responses.
        /// </summary>
        private SendDelivery mDelivery;

        /// <summary>
        /// Used for telling us to die.
        /// </summary>
        private volatile bool mSending = false;
        private volatile bool mDisposed = false;

        private SendPacket mSendPacket;
        private long mCursor;
        private long mTotal;
        private float mProgress;
        private bool mSendStatus;


        public SendDispatcher(Sender sender, SendDelivery delivery)
        {
            mQueue = new ConcurrentQueue<SendPacket>();
            mSender = sender;
            mDelivery = delivery;

            // Set Buffer
            SetBuffer(new byte[sender.GetBufferSize()], 0, sender.GetBufferSize());
        }

        public void Send(SendPacket entity)
        {
            mQueue.Enqueue(entity);

            if (!mSending)
            {
                mSending = true;

                // Send Next
                SendNext();
            }
        }

        private void SendAsync(int offset, int count)
        {
            if (mDisposed)
                return;

            if (count <= 0)
                SendNext();

            // Set Send Buffer Size
            SetBuffer(offset, count);

            // As soon as the client is connected, post a receive to the connection
            mSendStatus = mSender.SendAsync(this);
            if (!mSendStatus)
            {
                OnCompleted(this);
            }
        }

        private void SendNext()
        {
            // Set Status
            mSendStatus = mCursor == mTotal;


            SendPacket packet = mSendPacket;
            // Set Null
            mSendPacket = null;
            // Notity
            if (packet != null)
            {
                // End
                packet.EndPacket();

                // Post End
                SendDelivery delivery = mDelivery;
                if (delivery != null)
                {
                    packet.SetSuccess(mSendStatus);
                    delivery.PostSendProgress(packet, 1);
                }

                packet = null;
            }

            // Init Size
            mCursor = 0;
            mTotal = 0;

            // Take a request from the queue.
            mSending = mQueue.TryDequeue(out packet);

            if (mSending && packet != null)
            {
                // Cancel
                if (packet.IsCanceled())
                {
                    SendNext();
                }

                // Set Packet
                mSendPacket = packet;

                // Post Start
                SendDelivery delivery = mDelivery;
                if (delivery != null)
                    delivery.PostSendProgress(packet, 0);

                // Init the packet
                packet.StartPacket();

                // Send
                mSendStatus = SendHead(packet);

            }
        }

        private bool SendHead(SendPacket entity)
        {
            mTotal = entity.GetLength();
            if (mTotal <= 0)
                return false;

            // Type
            Buffer[0] = (byte)entity.GetType();

            // Length
            byte[] lenBytes = BitConverter.GetBytes(mTotal);
            lenBytes.CopyTo(Buffer, 1);

            // Info
            short infoLen = entity.ReadInfo(Buffer, HeadSize);
            byte[] infoLenBytes = BitConverter.GetBytes(infoLen);
            infoLenBytes.CopyTo(Buffer, HeadSize - 2);

            SendAsync(0, HeadSize + infoLen);

            return true;
        }

        private void SendEntity()
        {
            SendPacket packet = mSendPacket;
            if (packet != null)
            {
                int count = packet.Read(Buffer, 0, mSender.GetBufferSize());

                mCursor += count;

                // Send
                SendAsync(0, count);

                // Progress
                float progress = (float)Math.Round((float)mCursor / mTotal, 2);
                // Post Callback
                if (mProgress != progress)
                {
                    mProgress = progress;
                    SendDelivery delivery = mDelivery;
                    if (delivery != null)
                        delivery.PostSendProgress(packet, mProgress);
                }
            }
        }

        protected override void OnCompleted(SocketAsyncEventArgs e)
        {
            base.OnCompleted(e);

            // Check if the remote host closed the connection
            if (e.LastOperation == SocketAsyncOperation.Send
                && e.BytesTransferred > 0
                && e.SocketError == SocketError.Success)
            {
                if (e.BytesTransferred < e.Count)
                    SendAsync(e.Offset + e.BytesTransferred, e.Count - e.BytesTransferred);
                else if (mCursor != mTotal)
                    SendEntity();
                else
                    SendNext();
            }
            else
            {
                Dispose();
            }
        }

        public new void Dispose()
        {
            if (!mDisposed)
            {
                mDisposed = true;

                SendPacket packet = mSendPacket;
                mSendPacket = null;

                SendDelivery delivery = mDelivery;
                mDelivery = null;

                Sender sender = mSender;
                mSender = null;

                if (packet != null && delivery != null)
                {
                    if (mCursor < mTotal)
                    {
                        packet.EndPacket();
                        packet.SetSuccess(false);
                        delivery.PostSendProgress(packet, 1);
                    }
                }

                if (sender != null)
                    sender.Dispose();
                try
                {
                    SetBuffer(null, 0, 0);
                }
                catch (Exception) { }

                // Clear
                while (mQueue.TryDequeue(out packet)) { }

                base.Dispose();
            }

        }
    }
}
